File: pec-observation-app/ios-app/Sources/PECObservation/Application/UseCases/RecordAudioUseCase.swift
import Foundation

public final class RecordAudioUseCase {
    private let recorder: AudioRecordingPort
    private let repo: ObservationRepositoryPort
    private let syncQueue: SyncQueuePort

    public init(recorder: AudioRecordingPort, repo: ObservationRepositoryPort,
                syncQueue: SyncQueuePort) {
        self.recorder = recorder
        self.repo = repo
        self.syncQueue = syncQueue
    }

    public func startRecording(for observationId: UUID) async throws {
        guard var obs = try repo.fetchById(observationId) else {
            throw DomainError.notFound("Observation")
        }
        guard obs.isReadyForRecording else {
            throw DomainError.validationFailed("Preencha escola, professor, componente, série e tema antes de gravar")
        }

        let permitted = await recorder.requestPermission()
        guard permitted else { throw RecordingError.permissionDenied }

        obs.status = .readyToRecord
        try repo.save(obs)
        try await recorder.startRecording(observationId: observationId)
    }

    public func pauseRecording() async throws {
        try await recorder.pauseRecording()
    }

    public func resumeRecording() async throws {
        try await recorder.resumeRecording()
    }

    public func stopRecording(for observationId: UUID) async throws -> AudioRecordingResult {
        let result = try await recorder.stopRecording()

        guard var obs = try repo.fetchById(observationId) else {
            throw DomainError.notFound("Observation")
        }
        obs.audioRecording = result
        obs.status = .recorded
        obs.updatedAt = Date()
        try repo.save(obs)

        let command = SyncCommand(
            type: .uploadAudio,
            payload: [
                "observationId": observationId.uuidString,
                "audioLocalPath": result.localFileUrl.absoluteString,
                "checksum": result.checksum,
                "codec": result.codec,
                "durationSeconds": String(result.durationSeconds),
                "fileSizeBytes": String(result.fileSizeBytes),
            ]
        )
        try syncQueue.enqueue(command)
        return result
    }
}

File: pec-observation-app/android-app/app/src/main/kotlin/br/gov/educacao/sp/pecobservation/application/usecases/RecordAudioUseCase.kt
package br.gov.educacao.sp.pecobservation.application.usecases

import br.gov.educacao.sp.pecobservation.domain.entities.ObservationStatus
import br.gov.educacao.sp.pecobservation.domain.entities.SyncCommand
import br.gov.educacao.sp.pecobservation.domain.entities.SyncCommandType
import br.gov.educacao.sp.pecobservation.domain.errors.DomainError
import br.gov.educacao.sp.pecobservation.domain.errors.RecordingError
import br.gov.educacao.sp.pecobservation.ports.AudioRecordingPort
import br.gov.educacao.sp.pecobservation.ports.ObservationRepositoryPort
import br.gov.educacao.sp.pecobservation.ports.SyncQueuePort
import java.time.Instant
import java.util.UUID
import javax.inject.Inject

class RecordAudioUseCase @Inject constructor(
    private val recorder: AudioRecordingPort,
    private val repo: ObservationRepositoryPort,
    private val syncQueue: SyncQueuePort,
) {
    suspend fun startRecording(observationId: UUID) {
        val obs = repo.fetchById(observationId)
            ?: throw DomainError.NotFound("Observation", observationId.toString())

        if (!obs.isReadyForRecording)
            throw DomainError.ValidationError("observation", "not ready for recording")

        val permitted = recorder.requestPermission()
        if (!permitted) throw RecordingError.PermissionDenied()

        recorder.startRecording(observationId)
    }

    suspend fun stopAndSave(observationId: UUID) {
        val result = recorder.stopRecording()
        val obs = repo.fetchById(observationId)
            ?: throw DomainError.NotFound("Observation", observationId.toString())

        val updated = obs.copy(
            status         = ObservationStatus.RECORDED,
            audioRecording = result,
            updatedAt      = Instant.now(),
        )
        repo.save(updated)

        syncQueue.enqueue(
            SyncCommand(
                type = SyncCommandType.UPLOAD_AUDIO,
                payload = mapOf(
                    "observationId" to observationId.toString(),
                    "localFilePath" to result.localFilePath,
                    "checksum"      to result.checksum,
                    "durationSec"   to result.durationSeconds.toString(),
                    "fileSizeBytes" to result.fileSizeBytes.toString(),
                ),
            )
        )
    }
}
